### PR TITLE
Implement NFT functionality

### PR DESCRIPTION
**Pull request by bbyalcinkaya**
## Changes

* ESDT system SC
* Builtin funcions
  * `ESDTNFTCreate`
  * `ESDTNFTBurn`
  * `ESDTNFTAddQuantity`
* Host functions
  * `managedGetESDTTokenData`
  * `getESDTLocalRoles`
* Legacy async functions (required for token storage mappers)
* More composability tests

### CODE REVIEW

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md [hunk 1]
```diff
@@ -25,8 +25,14 @@ module MANDOS
         <newAddresses> .Map </newAddresses>
         <checkedAccounts> .Set </checkedAccounts>
         <elrond/>
+        <txCount> 0 </txCount>
         <exit-code exit=""> 0 </exit-code>
       </mandos>
+    
+    syntax Bytes ::= mkTxHash(Int)    [function, total]
+ // ---------------------------------------------------
+    rule mkTxHash(TX_CNT) => Int2Bytes(32, TX_CNT, BE)
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md:34]: Just curious: Is this our own implementation of creating a transaction hash, or is this also how Elrond implements it?

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/cryptoei.md [hunk 2]
```diff
@@ -48,6 +48,23 @@ module CRYPTOEI-HELPERS
                 ...
         </instrs>
 
+    syntax InternalInstr ::= "#verifyEd25519"   [klabel(verifyEd25519), symbol]
+ // ---------------------------------------------------------------------------
+
+    rule [verifyEd25519]:
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/cryptoei.md:54]: What is this supposed to do? Right now, it's a verification that always succeeds.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md [hunk 3]
```diff
@@ -221,5 +253,144 @@ module MANAGEDEI
...
+            <esdtMetadata> .esdtMetadata </esdtMetadata>
+            <esdtProperties> PROPS </esdtProperties>
+            ...
+          </esdtData>
+          ...
+        </account>
+
+    // extern long long getESDTLocalRoles(void* context, int32_t tokenIdHandle);
+    rule [getESDTLocalRoles]:
+        <instrs> hostCall("env", "getESDTLocalRoles", [ i32 .ValTypes ] -> [ i64 .ValTypes ])
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md:356]: How does this rule actually work? And why does it check that there are no esdtTransfers? Should it have a "#getESDTLocalRoles" on the RHS?

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/esdt.md [hunk 4]
```diff
@@ -90,6 +130,65 @@ module ESDT
...
+            <esdtBalance> ORIGFROM => ORIGFROM -Int DELTA </esdtBalance>
+            ...
+          </esdtData>
+          ...
+        </account>
+        <account>
+          <address> TO </address>
+          <esdtData>
+            <esdtId> TOKEN </esdtId>
+            <esdtBalance> ORIGFROM => ORIGFROM +Int DELTA </esdtBalance>
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/esdt.md:162]: s/ORIGFROM/ORIGTO/?

Also, what happens if FROM == TO?

Also, wouldn't it be simpler if this was split into an "removeFromSource" and "addToDestination"? The way it is written right now, the operation is called "addNFTToDestination", so its name implies either a mint operation, or the second half of a transfer, but it's actually a full transfer.

**Review by virgil-serbanuta — COMMENTED** [review]
You mentioned that this PR is not finished yet. Sorry if I commented about the unfinished stuff.

I didn't go through the entire code yet, I'll try to finish it a bit later.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md (hunk 3 above)
Comment (bbyalcinkaya) [reply kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md:356]: Sorry, this function is not finished yet. I should have excluded it in the commit.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/cryptoei.md (hunk 2 above)
Comment (bbyalcinkaya) [reply kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/cryptoei.md]: This was for a test I've decided not to include in this PR.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md [hunk 5]
```diff
@@ -102,6 +113,21 @@ Only take the next step once both the Elrond node and Wasm are done executing.
             #fi
         [simplification]
 
+    syntax MapBytesToBytes  ::= #removeReservedKeys ( MapBytesToBytes ) [function]
+ // ----------------------------------------------------------------------------------------
+    rule #removeReservedKeys(.MapBytesToBytes)
+        => .MapBytesToBytes
+    rule #removeReservedKeys(wrap(Key) Bytes2Bytes|-> Value M)
+        =>  #if #hasPrefix(Bytes2String(Key), "ELROND")
+            #then #removeReservedKeys(M)
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md:122]: this does not preserve definedness: if wrap(Key) is in M, then, on this branch, the LHS is not defined, while the RHS is. The same applies below.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md [hunk 6]
```diff
@@ -221,5 +253,145 @@ module MANAGEDEI
...
+          3 |-> <i32> VAL_IDX
+          4 |-> <i32> PROPS_IDX
+          ...
+        </locals>
+        <account>
+          <address> ADDR </address>
+          <esdtData>
+            <esdtId> TOKEN </esdtId>
+            <esdtBalance> VALUE </esdtBalance>
+            <esdtMetadata> .esdtMetadata </esdtMetadata>
```
Comment (virgil-serbanuta) [inline kmultiversx/src/kmultiversx/kdist/mx-semantics/vmhooks/managedei.md:347]: Why must the metadata be empty?

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md (hunk 5 above)
Comment (bbyalcinkaya) [reply kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md:122]: Should I use `keys_list(M)` and iterate over the list?

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/esdt.md [hunk 7]
```diff
@@ -316,6 +482,197 @@ module ESDT
...
+                                      <vmInput> 
+                                        <callValue> VALUE </callValue>
+                                        <callArgs> ARGS </callArgs>
+                                        _ 
+                                      </vmInput>)
+                => checkBool(VALUE ==Int 0, "built in function called with tx value is not allowed")
+                ~> checkBool(size(ARGS) >=Int 7, "invalid arguments to process built-in function")
+                ~> checkBool(SND ==K DST, "invalid receiver address")
+                ~> checkBool(lengthBytes(ARGS {{ 1 }} orDefault b"") <=Int 32,
+                      "invalid arguments to process built-in function, max length for quantity in nft create is 32")
```
[the remark this answers is not in the document]
Comment (bbyalcinkaya) [reply kmultiversx/src/kmultiversx/kdist/mx-semantics/esdt.md:505]: Yes, 32 is the limit for create NFT and add NFT quantity.

File: kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md (hunk 5 above)
Comment (virgil-serbanuta) [reply kmultiversx/src/kmultiversx/kdist/mx-semantics/mandos.md:122]: That would work. Another option would be to add `requires notBool wrap(Key) in M`

**Review by virgil-serbanuta — APPROVED** [review]

**Pull request merged**